File: src/HappyNumber.java
import java.util.HashSet;
import java.util.Set;

public class HappyNumber {
    public static boolean isHappy(int number) {
        return isHappy(number, new HashSet<>());
    }

    private static boolean isHappy(long number, Set<Long> sum) {
        long sumSquaresDigits = sumDigitSquares(number);
        if (sumSquaresDigits == 1) {
            return true;
        }
        if (sum.contains(sumSquaresDigits)) {
            return false;
        }
        sum.add(sumSquaresDigits);
        return isHappy(sumSquaresDigits, sum);
    }

    private static long sumDigitSquares(long number) {
        long result = 0;
        while (number > 0) {
            result += Math.pow(number % 10, 2);
            number /= 10;
        }
        return result;
    }
}

File: python/happy_number.py
class Solution:
    def _is_happy(self, number: int, previously_used: set) -> bool:
        if number is 1:
            return True
        new_number = sum(int(digit) ** 2 for digit in str(number))
        if new_number in previously_used:
            return False
        previously_used.add(new_number)
        return self._is_happy(new_number, previously_used)

    def isHappy(self, n: int) -> bool:
        return self._is_happy(number=n, previously_used={n})
